jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -78,14 +78,41 @@ class S3TreeDirectoryNode(private val bucket: S3VirtualBucket, parent: S3TreeDir
 
 private val fileTypeRegistry = FileTypeRegistry.getInstance()
 
-class S3TreeObjectNode(bucketName: String, parent: S3TreeDirectoryNode?, key: String, val size: Long, val lastModified: Instant) :
-    S3TreeNode(bucketName, parent, key) {
+open class S3TreeObjectNode(val bucket: S3VirtualBucket, parent: S3TreeDirectoryNode?, key: String, val size: Long, val lastModified: Instant) :
+    S3TreeNode(bucket.name, parent, key) {
 
+    var showHistory: Boolean = false
     private val fileType = fileTypeRegistry.getFileTypeByFileName(name)
 
     init {
         fileType.takeIf { it !is UnknownFileType }?.icon.let { icon = it }
     }
+
+    override fun getChildren(): Array<S3TreeNode> {
+        if (showHistory) {
+            val response = runBlocking {
+                bucket.listVersionObjects(key)
+            }
+            return (
+                response
+                    .versions()
+                    ?.map { S3TreeObjectVersionNode(bucket, parent, key, it.size(), it.lastModified(), it.versionId()) as S3TreeNode }
+                    ?: emptyList()
+                )
+                .toTypedArray()
+        }
+        return emptyArray()
+    }
+}
+
+class S3TreeObjectVersionNode(bucket: S3VirtualBucket, parent: S3TreeDirectoryNode?, key: String, size: Long, lastModified: Instant, val versionId: String) :
+    S3TreeObjectNode(bucket, parent, key, size, lastModified) {
+    override fun getName(): String = "${super.getName()}_$versionId"
+    private val fileType = fileTypeRegistry.getFileTypeByFileName(super.getName())
+
+    init {
```

Comment:
this `init` block shouldn't be needed since the constructor for `S3TreeObjectNode` will also run

Reply:
That would also make sense for me but when removed editor does not recognise file type anymore (look picture)
<img width="416" alt="Screenshot 2020-12-31 at 13 01 43" src="https://user-images.githubusercontent.com/6598894/103409858-f8560100-4b68-11eb-85a5-1127edb04c0b.png">
When I added it back the file type is recognised again.

Reply:
what might be happening is getName() returns `....txt_R0Xxah`, which the framework doesn't recognize as a text file?

try:
```kt
override fun getName(): String {
    val name = FileUtilRt.getNameWithoutExtension(super.getName())
    val originalExtension = FileUtilRt.getExtension(super.getName())
    val extension = if (originalExtension.isNotBlank()) ".$originalExtension" else ""

    return "$name_$versionId$extension"
}
```

Reply:
Neat!
<img width="762" alt="Screenshot 2021-01-02 at 10 20 47" src="https://user-images.githubusercontent.com/6598894/103454429-43485380-4ce4-11eb-9b0b-fc30f5740cdf.png">